### PR TITLE
feat: keyboard module improvements

### PR DESCRIPTION
**Pull request by eqyiel**
Hi @LnL7, thanks for your work on `nix-darwin`!

I wrote this so that I could do more flexible key bindings without resorting to Karabiner (which requries a kernel extension, which is a whole different kind of problem https://github.com/LnL7/nix-darwin/pull/91).

I'm using this module with this configuration:

```system.keyboard.enableKeyMapping = true;
system.keyboard.mappings = [
  {
    productId = 273;
    vendorId = 2131;
    mappings = {
      # Bind Caps Lock to Left Function for Realforce 87u.
      "Keyboard Caps Lock" = "Keyboard Left Function (fn)";
    };
  }

  {
    productId = 638;
    vendorId = 1452;
    mappings = {
      # For the built-in MacBook keyboard, change the modifiers to match a
      # traditional keyboard layout.
      "Keyboard Caps Lock" = "Keyboard Left Function (fn)";
      "Keyboard Left Alt" = "Keyboard Left GUI";
      "Keyboard Left Function (fn)" = "Keyboard Left Control";
      "Keyboard Left GUI" = "Keyboard Left Alt";
      "Keyboard Right Alt" = "Keyboard Right Control";
      "Keyboard Right GUI" = "Keyboard Right Alt";
    };
  }
];
```

Despite my best efforts there is a change in behaviour, and I changed it because I think it was wrong.

The `HIDKeyboardModifierMappingSrc` attribute set by `system.keyboard.nonUS.remapTilde` [was `30064771172`](https://github.com/LnL7/nix-darwin/blob/47fe6db9c9cd22c4ba57df3281f98f454880bea7/modules/system/keyboard.nix#L54), which is `Keyboard Non-US \ and |`, not `Keyboard Non-US # and ~`.

I changed it to reflect the name (it now remaps `"Keyboard Non-US # and ~" → "Keyboard Grave Accent and Tilde"` rather than `"Keyboard Non-US \ and |" → "Keyboard Grave Accent and Tilde"`).

I'm pretty sure that apart from that, there are no breaking changes.  If the old behaviour is preferred,  it's possible to get it back like this:

```
system.keyboard.enableKeyMapping = true;
system.keyboard.mappings = {
  "Keyboard Non-US \ and |" = "Keyboard Grave Accent and Tilde";
};
```

I also had to change the test, becuase it's not possible for Caps Lock to be bound to both Left Control and Escape at the same time.  Previously, it passed two `hidutil` commands would be in the activation script: one to bind Caps Lock to Left Control, and another to bind Caps Lock to Escape (which overwrote the first).

<details>
<summary>Mapped HID usage table for reference</summary>

```
# lib.mapAttrs (name: value: builtins.bitOr 30064771072 value) (import ./hid-usage-table.nix)

{
  "Keyboard  and |" = 30064771121;
  "Keyboard ' and \"" = 30064771124;
  "Keyboard , and \"<\"" = 30064771126;
  "Keyboard - and _" = 30064771117;
  "Keyboard . and \">\"" = 30064771127;
  "Keyboard / and ?" = 30064771128;
  "Keyboard 0 and )" = 30064771111;
  "Keyboard 1 and !" = 30064771102;
  "Keyboard 2 and @" = 30064771103;
  "Keyboard 3 and #" = 30064771104;
  "Keyboard 4 and $" = 30064771105;
  "Keyboard 5 and %" = 30064771106;
  "Keyboard 6 and ^" = 30064771107;
  "Keyboard 7 and &" = 30064771108;
  "Keyboard 8 and *" = 30064771109;
  "Keyboard 9 and (" = 30064771110;
  "Keyboard ; and :" = 30064771123;
  "Keyboard = and +" = 30064771118;
  "Keyboard Application" = 30064771173;
  "Keyboard Caps Lock" = 30064771129;
  "Keyboard Delete (Backspace)" = 30064771114;
  "Keyboard Delete Forward" = 30064771148;
  "Keyboard Down Arrow" = 30064771153;
  "Keyboard End" = 30064771149;
  "Keyboard Escape" = 30064771113;
  "Keyboard F1" = 30064771130;
  "Keyboard F10" = 30064771139;
  "Keyboard F11" = 30064771140;
  "Keyboard F12" = 30064771141;
  "Keyboard F13" = 30064771176;
  "Keyboard F14" = 30064771177;
  "Keyboard F15" = 30064771178;
  "Keyboard F16" = 30064771179;
  "Keyboard F17" = 30064771180;
  "Keyboard F18" = 30064771181;
  "Keyboard F19" = 30064771182;
  "Keyboard F2" = 30064771131;
  "Keyboard F20" = 30064771183;
  "Keyboard F21" = 30064771184;
  "Keyboard F22" = 30064771185;
  "Keyboard F23" = 30064771186;
  "Keyboard F24" = 30064771187;
  "Keyboard F3" = 30064771132;
  "Keyboard F4" = 30064771133;
  "Keyboard F5" = 30064771134;
  "Keyboard F6" = 30064771135;
  "Keyboard F7" = 30064771136;
  "Keyboard F8" = 30064771137;
  "Keyboard F9" = 30064771138;
  "Keyboard Grave Accent and Tilde" = 30064771125;
  "Keyboard Home" = 30064771146;
  "Keyboard Insert" = 30064771145;
  "Keyboard Left Alt" = 30064771298;
  "Keyboard Left Arrow" = 30064771152;
  "Keyboard Left Control" = 30064771296;
  "Keyboard Left GUI" = 30064771299;
  "Keyboard Left Shift" = 30064771297;
  "Keyboard Non-US  and |" = 30064771172;
  "Keyboard Non-US # and ~" = 30064771122;
  "Keyboard Page Down" = 30064771150;
  "Keyboard Page Up" = 30064771147;
  "Keyboard Pause" = 30064771144;
  "Keyboard Power" = 30064771174;
  "Keyboard Print Screen" = 30064771142;
  "Keyboard Return (Enter)" = 30064771112;
  "Keyboard Right Alt" = 30064771302;
  "Keyboard Right Arrow" = 30064771151;
  "Keyboard Right Control" = 30064771300;
  "Keyboard Right GUI" = 30064771303;
  "Keyboard Right Shift" = 30064771301;
  "Keyboard Scroll Lock" = 30064771143;
  "Keyboard Spacebar" = 30064771116;
  "Keyboard Tab" = 30064771115;
  "Keyboard Up Arrow" = 30064771154;
  "Keyboard [ and {" = 30064771119;
  "Keyboard ] and }" = 30064771120;
  "Keyboard a and A" = 30064771076;
  "Keyboard b and B" = 30064771077;
  "Keyboard c and C" = 30064771078;
  "Keyboard d and D" = 30064771079;
  "Keyboard e and E" = 30064771080;
  "Keyboard f and F" = 30064771081;
  "Keyboard g and G" = 30064771082;
  "Keyboard h and H" = 30064771083;
  "Keyboard i and I" = 30064771084;
  "Keyboard j and J" = 30064771085;
  "Keyboard k and K" = 30064771086;
  "Keyboard l and L" = 30064771087;
  "Keyboard m and M" = 30064771088;
  "Keyboard n and N" = 30064771089;
  "Keyboard o and O" = 30064771090;
  "Keyboard p and P" = 30064771091;
  "Keyboard q and Q" = 30064771092;
  "Keyboard r and R" = 30064771093;
  "Keyboard s and S" = 30064771094;
  "Keyboard t and T" = 30064771095;
  "Keyboard u and U" = 30064771096;
  "Keyboard v and V" = 30064771097;
  "Keyboard w and W" = 30064771098;
  "Keyboard x and X" = 30064771099;
  "Keyboard y and Y" = 30064771100;
  "Keyboard z and Z" = 30064771101;
  "Keypad *" = 30064771157;
  "Keypad +" = 30064771159;
  "Keypad -" = 30064771158;
  "Keypad . and Delete" = 30064771171;
  "Keypad /" = 30064771156;
  "Keypad 0 and Insert" = 30064771170;
  "Keypad 1 and End" = 30064771161;
  "Keypad 2 and Down Arrow" = 30064771162;
  "Keypad 3 and Page Down" = 30064771163;
  "Keypad 4 and Left Arrow" = 30064771164;
  "Keypad 5" = 30064771165;
  "Keypad 6 and Right Arrow" = 30064771166;
  "Keypad 7 and Home" = 30064771167;
  "Keypad 8 and Up Arrow" = 30064771168;
  "Keypad 9 and Page Up" = 30064771169;
  "Keypad =" = 30064771175;
  "Keypad Enter" = 30064771160;
  "Keypad Num Lock and Clear" = 30064771155;
}

```

</details>

### CODE REVIEW

File: release.nix [hunk 1]
```diff
@@ -72,7 +72,8 @@ let
     unstable = pkgs.releaseTools.aggregate {
       name = "darwin-${pkgs.lib.nixpkgsVersion}";
       constituents =
-        [ jobs.stdenv.x86_64-darwin
+        [
+          jobs.stdenv.x86_64-darwin
```
Comment (eqyiel) [inline release.nix:76]: Ah, sorry, this is due to `nixpkgs-fmt`, I'll amend it to remove the unnecessary diff

**Comment by eqyiel** [comment]
I noticed that `/Library/LaunchDaemons/org.nixos.activate-system.plist` doesn't get applied after waking from sleep, meaning that the bindings vanish 😢 I guess the keyboard activation part should be moved a `LaunchAgent` instead

**Comment by eqyiel** [comment]
> I noticed that /Library/LaunchDaemons/org.nixos.activate-system.plist doesn't get applied after waking from sleep, meaning that the bindings vanish

I solved this using the "LaunchEvents" feature, but it requires using [a helper program](https://github.com/snosrap/xpc_set_event_stream_handler) to mark the event as consumed (otherwise the script will be retried forever).  I'm not 100% sure about adding packages to this repo, but I'd like to hear your opinion about this @LnL7 🙏 

I'm also not happy about the retry/timeout thing, but I couldn't find another way to get `launchd` to retry the script until it exits successfully.

File: modules/system/keyboard/default.nix [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+                IOMatchLaunchStream = true;
+                IOProviderClass = "IOUSBDevice";
+                idProduct = productId;
+                idVendor = vendorId;
+              };
+            };
+          };
+          serviceConfig.RunAtLoad = true;
+        })
+        ))
```
Comment (eqyiel) [inline modules/system/keyboard/default.nix:296]: IMO this is strange formatting but it's what nixpkgs-fmt does and I don't care enough to stop using it

**Comment by eqyiel** [comment]
I ended up making this a local module 😊

**Comment by nathanshelly** [comment]
Would you mind sharing that local module? This feature sounds really useful so would love to have it in my own setup

**Comment by behaghel** [comment]
@eqyiel may I ask why did you close this one? Seems good and useful work to me.

**Comment by eqyiel** [comment]
@nathanshelly @behaghel here's a gist https://gist.github.com/eqyiel/412bc7acb68d1d15826893989f84fca7, there is one dependency: https://github.com/LnL7/nix-darwin/pull/455

I closed it because it had been more than a year 😞 I get that people are busy, I'm busy too, personally I still use this but I don't want to keep maintaining this PR.

**Pull request closed without merging**